cpp/src/io/json/read_json.cu

```diff
@@ -278,25 +306,51 @@ table_with_metadata read_json(host_span<std::unique_ptr<datasource>> sources,
                  "Multiple inputs are supported only for JSON Lines format");
   }
 
-  datasource::owning_buffer<rmm::device_uvector<char>> bufview =
-    get_record_range_raw_input(sources, reader_opts, stream);
+  std::for_each(sources.begin(), sources.end(), [](const auto& source) {
+    CUDF_EXPECTS(source->size() < std::numeric_limits<int>::max(),
+                 "The size of each source file must be less than INT_MAX bytes");
+  });
 
-  // If input JSON buffer has single quotes and option to normalize single quotes is enabled,
-  // invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_single_quotes()) {
-    normalize_single_quotes(bufview, stream, rmm::mr::get_current_device_resource());
-  }
+  size_t const batch_size = std::numeric_limits<int>::max();
+  int const num_batches   = std::ceil((double)sources_size(sources, 0, 0) / batch_size);
+  if (num_batches == 1) return read_batch(sources, reader_opts, stream, mr);
 
-  // If input JSON buffer has unquoted spaces and tabs and option to normalize whitespaces is
-  // enabled, invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_whitespace()) {
-    normalize_whitespace(bufview, stream, rmm::mr::get_current_device_resource());
-  }
+  std::vector<size_t> prefsum_source_sizes(sources.size());
+  std::transform_inclusive_scan(sources.begin(),
+                                sources.end(),
+                                prefsum_source_sizes.begin(),
+                                std::plus<size_t>{},
+                                [](const std::unique_ptr<datasource>& s) { return s->size(); });
 
-  auto buffer =
-    cudf::device_span<char const>(reinterpret_cast<char const*>(bufview.data()), bufview.size());
-  stream.synchronize();
-  return device_parse_nested_json(buffer, reader_opts, stream, mr);
+  auto start_source_it = sources.begin();
+  std::vector<cudf::io::table_with_metadata> partial_tables;
+  for (int batch = 0; batch < num_batches; batch++) {
+    auto end_prefsum_it = std::upper_bound(
+      prefsum_source_sizes.begin(), prefsum_source_sizes.end(), (batch + 1) * batch_size);
+    auto end_source_it =
+      sources.begin() + std::distance(prefsum_source_sizes.begin(), end_prefsum_it);
+    partial_tables.emplace_back(
+      read_batch(host_span<std::unique_ptr<datasource>>(
+                   start_source_it, std::distance(start_source_it, end_source_it)),
+                 reader_opts,
```

Comment:
I don't see how this works with byte_range. Seems like we would read the given byte range from every batch, instead of reading it once from the entire input.

Reply:
Thanks for pointing this out - I have modified the byte range parsing to be in batches as well.